src/common/pkcs11/pkcs11.cpp

```diff
@@ -0,0 +1,1077 @@
+    }
+
+    dst.Resize(size);
+
+    return ErrorEnum::eNone;
+}
+
+template <size_t cSize>
+Error Convert(const CK_UTF8CHAR (&src)[cSize], String& dst)
+{
+    return Convert(Array<uint8_t>(src, cSize), dst);
+}
+
+template <size_t cSize>
+Error Convert(const String& src, CK_UTF8CHAR (&dst)[cSize])
+{
+    if (cSize <= src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    memset(dst, ' ', cSize);
+    memcpy(dst, src.CStr(), src.Size());
+
+    return ErrorEnum::eNone;
+}
+
+void Convert(const CK_VERSION& src, Version& dst)
+{
+    dst.mMajor = src.major;
+    dst.mMinor = src.minor;
+}
+
+Error Convert(const CK_SLOT_INFO& src, SlotInfo& dst)
+{
+    Error err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.slotDescription, dst.mSlotDescription);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    Convert(src.hardwareVersion, dst.mHardwareVersion);
+    Convert(src.firmwareVersion, dst.mFirmwareVersion);
+
+    dst.mFlags = src.flags;
+
+    return ErrorEnum::eNone;
+}
+
+Error Convert(const CK_TOKEN_INFO& src, TokenInfo& dst)
+{
+    Error err = Convert(src.label, dst.mLabel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.model, dst.mModel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.serialNumber, dst.mSerialNumber);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    dst.mFlags = src.flags;
+    Convert(src.hardwareVersion, dst.mHardwareVersion);
+    Convert(src.firmwareVersion, dst.mFirmwareVersion);
+    dst.mTotalPublicMemory = src.ulTotalPublicMemory;
+    dst.mFreePublicMemory = src.ulFreePublicMemory;
+    dst.mTotalPrivateMemory = src.ulTotalPrivateMemory;
+    dst.mFreePrivateMemory = src.ulFreePrivateMemory;
+
+    return ErrorEnum::eNone;
+}
+
+Error Convert(const CK_INFO& src, LibInfo& dst)
+{
+    Convert(src.cryptokiVersion, dst.mCryptokiVersion);
+    Convert(src.libraryVersion, dst.mLibraryVersion);
+
+    Error err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.libraryDescription, dst.mLibraryDescription);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Convert(const Array<ObjectAttribute>& src, Array<CK_ATTRIBUTE>& dst)
+{
+    if (dst.MaxSize() < src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    dst.Clear();
+
+    for (const auto& attr : src) {
+        CK_ATTRIBUTE tmp;
+
+        tmp.type = attr.mType;
+        tmp.pValue = (void*)attr.mValue.Get();
+        tmp.ulValueLen = attr.mValue.Size();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error BuildAttributes(const Array<AttributeType>& types, Array<Array<uint8_t>>& values, Array<CK_ATTRIBUTE>& dst)
```

Comment:
[al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433761567)

Merge what? Provide more meaningful name.
Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433933880)

CK_ATTRIBUTE consists of type and value
this method merges two distinct type and value arrays into a single CK_ATTRIBUTE array.

I will renames it to BuildAttributes.